providers/digitalocean/digitaloceanProvider.go

```diff
@@ -0,0 +1,197 @@
+		context.Background(),
+		oauth2.StaticTokenSource(&oauth2.Token{AccessToken: m["token"]}),
+	)
+	client := godo.NewClient(oauthClient)
+
+	return &DoApi{client: client}, nil
+}
+
+func init() {
+	providers.RegisterDomainServiceProviderType("DIGITALOCEAN", newDo, providers.CanUseSRV)
+}
+
+func (api *DoApi) EnsureDomain(domain string) error {
+	ctx := context.Background()
+	_, resp, err := api.client.Domains.Get(ctx, domain)
+	if resp.Status == "404" {
+		_, _, err := api.client.Domains.Create(ctx, &godo.DomainCreateRequest{
+			Name:      domain,
+			IPAddress: "",
+		})
+		return err
+	} else {
+		return err
+	}
+}
+
+func (api *DoApi) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return models.StringsToNameservers(defaultNameServerNames), nil
+}
+
+func (api *DoApi) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	ctx := context.Background()
+	dc.Punycode()
+
+	records := []godo.DomainRecord{}
+	opt := &godo.ListOptions{}
+	for {
+		result, resp, err := api.client.Domains.Records(ctx, dc.Name, opt)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, d := range result {
+			records = append(records, d)
+		}
+
+		if resp.Links == nil || resp.Links.IsLastPage() {
+			break
+		}
+
+		page, err := resp.Links.CurrentPage()
+		if err != nil {
+			return nil, err
+		}
+
+		opt.Page = page + 1
+	}
+
+	existingRecords := make([]*models.RecordConfig, len(records))
+	for i := range records {
+		existingRecords[i] = toRc(dc, &records[i])
+	}
+
+	differ := diff.New(dc)
+	_, create, delete, modify := differ.IncrementalDiff(existingRecords)
+
+	var corrections = []*models.Correction{}
+
+	// Deletes first so changing type works etc.
+	for _, m := range delete {
+		id := m.Existing.Original.(*godo.DomainRecord).ID
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s, DO ID: %d", m.String(), id),
+			F: func() error {
+				_, err := api.client.Domains.DeleteRecord(ctx, dc.Name, id)
+				return err
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+	for _, m := range create {
+		req := toReq(dc, m.Desired)
+		corr := &models.Correction{
+			Msg: m.String(),
+			F: func() error {
+				_, _, err := api.client.Domains.CreateRecord(ctx, dc.Name, req)
+				return err
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+	for _, m := range modify {
+		id := m.Existing.Original.(*godo.DomainRecord).ID
+		req := toReq(dc, m.Desired)
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s, DO ID: %d", m.String(), id),
+			F: func() error {
+				_, _, err := api.client.Domains.EditRecord(ctx, dc.Name, id, req)
+				return err
+			},
+		}
+		corrections = append(corrections, corr)
+	}
+
+	return corrections, nil
+}
+
+func toRc(dc *models.DomainConfig, r *godo.DomainRecord) *models.RecordConfig {
+	// This handles "@" etc.
+	name := dnsutil.AddOrigin(r.Name, dc.Name)
+
+	target := r.Data
+	// Make target FQDN (#rtype_variations)
+	if r.Type == "CNAME" && target == "@" {
```

Comment:
I'm not quite sure on the logic here. Is the first case for something like `CNAME('foo', '@')`? I'm not sure I've considered that very often. Why are the other record types not included in the first if?

I feel like maybe that is something worth adding an integration test for.